automake/post/rules/makedirs.am

```diff
@@ -50,23 +63,72 @@ $1: | makedirs
 endef # makedirs-depends
 
 # Use brute force: before any compilation, all _OBJECTS depend on makedirs
+#  wish there were something in automake that meant "please do first"...
 $(foreach objects,$(filter %_OBJECTS,$(.VARIABLES)),\
    $(foreach object,$($(objects)),\
    $(eval $(call makedirs-depends,$(object)))))
 
-makedirs: force
-	$(foreach dir,$(MAKEDIRS),$(call makedirs-make,$(dir)))
+# This says chunk says:
+#  1. makedirs (the order-only PHONY target above) depends on the file
+#       makedirs_id in each of the MAKEDIR directories
+#  2. build $(makedir_id) by running make, then touching the id file
+#  3. clean and distclean should whack all makedir_id files that might
+#       have accumulated
+define makedirs-makedir-rules
+makedirs: $1/$(makedirs_id)
+
+$1/$(makedirs_id):
+	$(NL_V_AT)$(MAKE) -C $1
+	$(NL_V_AT)touch $$@
+	$(NL_V_AT)$(RM) $$(filter-out $$@,$$(wildcard $1/makedirs_*))
+
+$1/makedirs_clean:
+	$(NL_V_AT)$(RM) $$(wildcard $1/makedirs_*)
+
+.PHONY: $1/makedirs_clean
+distclean clean: $1/makedirs_clean
+
+endef # define makedirs-makedir-rules
+
+# run the above chunk for each MAKEDIR
+$(foreach dir,$(MAKEDIRS),\
+	$(eval $(call makedirs-makedir-rules,$(dir))))
+
 
 endif # ifdef MAKEDIRS
 
+
+#
+# MAKEDIR_TARGETS section
+#
 ifdef MAKEDIR_TARGETS
-# The way me make a MAKEDIR_TARGET
-define makedirs-makedir-target-rule
-$1: force
-	$(MAKE) -C $$(dir $$(@)) $$(notdir $$(@))
-endef # makedirs-makedir-target-rule
 
+# This chunk says:
+#  1. the target ($1 in this function) depends on the target's
+#       makedir_id file in its directory
+#  2. build the target with make
+#  2. build the target's makedir_id by touching the id file (and
+#       whacking any leftovers, accumulated ones)
+#  3. clean and distclean should whack all makedir_id files that might
+#       have accumulated
+define makedirs-makedir-target-rules
+$1: $(notdir $1).$(makedirs_id)
```

Comment:
This tripped me up in terms of the sentinel placement without the explicit path separator, but I now see that this is, in fact, in the target subdirectory.

Reply:
yeah, but we should keep the target in the name of the sentinel in case there's a directory with more than one useful product, fixed in my latest.